### PR TITLE
Release v2.0

### LINKED ISSUE #88: Receiving warning: oasOpParams expects a resolved object
### What is the current behavior?

I am currently migrating my API specification and running into this warning:

```
oasOpParams expects a resolved object, but none was provided. Results may not be correct.
oasPathParam expects a resolved object, but none was provided. Results may not be correct.
```

There are no `$ref` mentions in my spec, so I am confused as to why I am getting this error. The code is as follows.

### What is the expected behavior?

For this not to error

### Please tell us about your environment:

  - Version:   `@stoplight/spectral": "^1.0.1"`
  - Framework: Node v11.7.0
  - Language: Node v11.7.0

### Other information

My code:

```js
// Load Spectral and its OAS3 validations
const { Spectral } = require('@stoplight/spectral')
const { oas3Functions, oas3Rules } = require('@stoplight/spectral/rulesets/oas3')

// Find all OpenAPI files
const fs = require('fs')
const glob = require('glob')

const lint = async () => {
  const fileNames = await new Promise((resolve, reject) => {
    glob('./v2.0/*.openapi.json', {}, (error, match) => {
      if (error) { reject(error) }
      else { resolve(match) }
    })
  })
  
  // Initialize Spectral
  const spectral = new Spectral()
  spectral.addFunctions(oas3Functions())
  spectral.addRules(oas3Rules())
  
  fileNames.forEach(fileName => {
    let specification = JSON.parse(fs.readFileSync(fileName))
    const results = spectral.run(specification)
    // console.dir(results)
  })  
}

lint()
```

### LINKED ISSUE #91: Spectral is generating paths that do not exist in the original document
It seems like the included rules for OAS2/3 can generate paths that do not exist in the original document.

An easy way to reproduce this is to simply install the default rules and create an OAS document with a path that's missing the `description` property.

### LINKED ISSUE #92: Spectral's IRuleResult an IPosition reference which is always null
Spectral's `IRuleResult` interface which is the type returned from the validation process extends `IValidationResult` which has a nullable position reference.

While it is correct — Spectral does not have (yet?) the ability of generating the position information — these are up to the caller.

In order to reduce the confusion (which I personally had) it would be better to simply make sure such interface does not have that property at all.

### LINKED ISSUE #109: Spectral OAS2 validation throws a non-existent error without path value
Running spectral OAS2 rules and functions sets on a OAS2 `.yaml` file on all tests (2 different `yaml files`) results in a cryptic error without a `"path"` parameter:
```
{
    "name": "oas2-schema",
    "summary": "Validate structure of OpenAPIv2 specification.",
    "message": "should NOT have additional properties: 0",
    "path": [],
    "severity": 50,
    "severityLabel": "error"
}
```
Those files don't throw any errors in any other validation tool (swagger editor, apiary, etc.). I am happy to provide them if needed.

### **I'm submitting a...**
  - [X] bug report
  - [ ] feature request

### What is the current behavior?

Running:
```
spectral.addFunctions(oas2Functions());
spectral.addRules(oas2Rules());
const results = spectral.run(myOAS);
```
on a OAS2 `.yaml` file

### What is the expected behavior?
Do not throw this error or be more specific on it's origin and place in the file.

### What is the motivation / use case for changing the behavior?
User experience and compliance with OAS2 specification

### Please tell us about your environment:

  - Version: @stoplight/spectral@1.2.0
  - Framework: [ ]
  - Language: [all]

### Other information

It also throws 
```
oasOpParams expects a resolved object, but none was provided. Results may not be correct.
oasPathParam expects a resolved object, but none was provided. Results may not be correct.
```
which does not say anything of value without more explanation.

### PR DESCRIPTION
**Pull request by philsturgeon**
### Please check if the PR fulfills these requirements
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce?

A whole bunch of new features, and a few breaking changes. We have a new CLI with two formatters to start off with (stylish and json) and the JS library interface has been simplified to no longer need resolvedTarget.

### Does this PR introduce a breaking change?

If you are passing resolvedTarget, take that out. We will automatically resolve whatever you pass in using https://github.com/stoplightio/json-ref-resolver. Future versions will document how to pass in your own resolcvers once we have figured out exactly how to approach it. 

### Other information

fix #88
fix #91
fix #92
fix #109

### CODE REVIEW

**Comment by philsturgeon** [comment]
Hey @MikeRalphson, fancy taking a quick look at this beast? We'll be hitting that big beautiful green button later today.

File: README.md [hunk 1]
```diff
@@ -24,14 +24,86 @@ Supports Node v8.3+ and modern browsers.
...
+
+Spectral can be run via the command-line:
+
+```bash
+spectral lint petstore.yaml
+```
+
+Other options include:
+
+```bash
```
Comment (MikeRalphson) [inline README.md]: Use of `bash` here results in 'random' highlighting of keywords such as `for` and `help`.

File: src/__tests__/functions.test.ts [hunk 2]
```diff
@@ -53,11 +53,11 @@ describe('functions', () => {
...
             },
           }
-        ).length
-      ).toEqual(1);
+        )
+      ).resolves.toHaveLength(1);
     });
 
-    test('dont returns results if one of the properties are present', () => {
+    test('dont returns results if one of the properties are present', async () => {
```
Comment (MikeRalphson) [inline src/__tests__/functions.test.ts]: "dont return" or "doesn't return"?

File: src/__tests__/linter.test.ts [hunk 3]
```diff
@@ -30,20 +30,20 @@ const rules = {
 describe('linter', () => {
   const spectral = new Spectral();
 
-  test('should not lint if passing in value is not an object', () => {
+  test('should not lint if passing in value is not an object', async () => {
```
Comment (MikeRalphson) [inline src/__tests__/linter.test.ts]: "passed in object".

File: src/rulesets/oas/__tests__/__snapshots__/path-not-include-query.ts.snap [hunk 4]
```diff
@@ -3,14 +3,24 @@
 exports[`path-not-include-query return errors if indlues a query 1`] = `
```
Comment (MikeRalphson) [inline src/rulesets/oas/__tests__/__snapshots__/path-not-include-query.ts.snap]: "includes".

File: src/rulesets/oas/__tests__/operation-2xx-response.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    expect(results).toEqual([
+      expect.objectContaining({
+        code: 'operation-2xx-response',
+        message: 'operations must define at least one 2xx response',
+        path: ['paths', '/path', 'get', 'responses'],
+      }),
+    ]);
+  });
+
+  test('ignores any at the PathItem level which is not a HTTP verb', async () => {
```
Comment (MikeRalphson) [inline src/rulesets/oas/__tests__/operation-2xx-response.ts]: Is this title missing "vendor extensions", and possibly "verb" -> "method".

File: README.md [hunk 6]
```diff
@@ -24,14 +24,86 @@ Supports Node v8.3+ and modern browsers.
 
 ## Usage
 
+### CLI
+
+Spectral can be run via the command-line:
+
+```bash
+spectral lint petstore.yaml
```
Comment (MikeRalphson) [inline README.md]: As we don't instruct users to install globally, is this technically correct?

Comment (philsturgeon) [reply README.md]: We are adding this in https://github.com/stoplightio/spectral/pull/113

**Comment by MikeRalphson** [comment]
Linting this document:

```yaml
openapi: 3.0.8
info:
  title: occaecat ex exercitation consequat
  version: sunt Duis aliqua qui
x-testcase: 'should not work without required property: paths'
```

Results in stylish format output:

```
linting ../openapi3-examples/fail/fuzz1/331be1bf-781d-407f-93d6-1f4b390ae32b.yaml
OpenAPI 3.x detected

/home/mike/node/openapi3-examples/fail/fuzz1/331be1bf-781d-407f-93d6-1f4b390ae32b.yaml
 1:1    error  oas3-schema       Validate structure of OpenAPIv3 specification
 1:1  warning  api-servers       OpenAPI `servers` must be present and non-empty array
 2:6  warning  info-contact      Info object should contain `contact` object
 2:6  warning  info-description  OpenAPI object info `description` must be present and non-empty string

✖ 4 problems (1 error, 3 warnings, 0 infos)
```

But the json format output gives the more helpful message for the error:

```json
	{
		"code": "oas3-schema",
		"path": [],
		"message": "should have required property 'paths'",
		"severity": 0
	},
```

**Comment by philsturgeon** [comment]
Yeah this is a tough one. Most of the time the summary is a bit more concise but for oasX-schema based rules you really want the message. Id rather not make the formatters be aware of such things, so we need to find somewhere slightly higher up to make that change perhaps.

**Comment by philsturgeon** [comment]
@MikeRalphson I moved this to #116 as it does not need to block v2.0.

File: src/formatters/json.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { IRuleResult } from '../types';
+
+export const json = (results: IRuleResult[]): string => {
+  const outputJson = results.map(result => {
+    return {
+      code: result.code,
+      path: result.path,
+      message: result.message,
+      severity: result.severity,
```
[the remark this answers is not in the document]
Comment (P0lip) [reply src/formatters/json.ts]: I'll add it right away.

**Pull request merged**